contracts/ExponentialStaking.sol

```diff
@@ -0,0 +1,267 @@
+            asset.transferFrom(msg.sender, address(this), amountIn);
+            // amountIn already added into newAmount during initialization
+        }
+
+        // Collect funds from old stake (optional)
+        if (lockupId != NEW_STAKE) {
+            lockup = lockups[to][uint256(lockupId)];
+            uint256 oldAmount = lockup.amount;
+            oldEnd = lockup.end;
+            oldPoints = lockup.points;
+            require(oldAmount > 1, "Staking: Already closed stake");
+            emit Unstake(to, uint256(lockupId), oldAmount, oldEnd, oldPoints);
+            newAmount += oldAmount;
+        }
+
+        // Collect funds from rewards (optional)
+        newAmount += _collectRewards(to, stakeRewards);
+
+        // Caculate Points and lockup
+        require(newAmount > 0, "Staking: Not enough");
+        require(newAmount <= type(uint128).max, "Staking: Too much");
+        (uint256 newPoints, uint256 newEnd) = previewPoints(newAmount, duration);
+        require(newPoints + totalSupply() <= type(uint192).max, "Staking: Max points exceeded");
+        lockup.end = uint128(newEnd);
+        lockup.amount = uint128(newAmount); // max checked in require above
+        lockup.points = newPoints;
+
+        // Update or create lockup
+        if (lockupId != NEW_STAKE) {
+            require(newEnd > oldEnd, "Staking: New lockup must be longer");
+            lockups[to][uint256(lockupId)] = lockup;
+        } else {
+            lockups[to].push(lockup);
+            uint256 numLockups = lockups[to].length;
+            // Delegate voting power to the receiver, if unregistered and first stake
+            if (numLockups == 1 && delegates(to) == address(0)) {
+                _delegate(to, to);
+            }
+            require(numLockups < uint256(type(int256).max), "Staking: Too many lockups");
+        }
+        _mint(to, newPoints - oldPoints);
+        emit Stake(to, uint256(lockupId), newAmount, newEnd, newPoints);
+    }
+
+    /// @notice Collect staked asset for a lockup and any earned rewards.
+    /// @param lockupId the id of the lockup to unstake
+    function unstake(uint256 lockupId) external {
+        Lockup memory lockup = lockups[msg.sender][lockupId];
+        uint256 amount = lockup.amount;
+        uint256 end = lockup.end;
+        uint256 points = lockup.points;
+        require(end != 0, "Staking: Already unstaked this lockup");
+        _collectRewards(msg.sender, false);
+
+        uint256 withdrawAmount = previewWithdraw(amount, end);
+        uint256 penalty = amount - withdrawAmount;
+
+        delete lockups[msg.sender][lockupId]; // Keeps empty in array, so indexes are stable
+        _burn(msg.sender, points);
+        if (penalty > 0) {
+            asset.transfer(address(rewardsSource), penalty);
+        }
+        asset.transfer(msg.sender, withdrawAmount);
+        emit Unstake(msg.sender, lockupId, withdrawAmount, end, points);
```

Comment:
Wonder if we should add the penalty to the event to make it easier for:
1. analytics
2. off-chain calculation of how much total penalties occurred so that it can be factored in for updating the rate emission in the future (since penalty should be redistributed as extra rewards to other stakers).